File: artifacts/product-reality/sprint-191/m05/verify-census.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import {createHash} from 'node:crypto';
const base='artifacts/product-reality/sprint-191/m05';
const read=file=>JSON.parse(fs.readFileSync(file,'utf8'));
const hash=bytes=>createHash('sha256').update(bytes).digest('hex');
const before=read('artifacts/product-reality/sprint-190/m06/final/proof/component-census/report.json'),after=read(`${base}/component-census/report.json`);
const key=row=>`${row.input.object}/${row.input.context}`;
const changes=[],statusChanges=[];
assert.equal(after.greenTotalSchemas,77);assert.equal(after.greenTotalCells,154);
for(const row of after.allRows) {
 const old=before.allRows.find(prior=>key(prior)===key(row));assert(old);
 if(old.green!==row.green) statusChanges.push(key(row));
 for(const cell of row.cells) {
  const previous=old.cells.find(candidate=>candidate.framework===cell.framework);
  const first=read(previous.response.path).artifact,last=read(cell.response.path).artifact;
  if(first?.contentHash===last?.contentHash) continue;
  const files=last.files.filter(file=>!first?.files.some(old=>old.path===file.path&&old.contentHash===file.contentHash)).map(file=>({file:file.path,beforeSha256:first?.files.find(old=>old.path===file.path)?.contentHash??null,afterSha256:file.contentHash}));
  changes.push({schema:key(row),framework:cell.framework,beforeArtifact:first?.contentHash??null,afterArtifact:last.contentHash,files});
 }
}
assert.deepEqual(statusChanges.sort(),['Organization/workflow','User/workflow']);
const workflowCells=after.allRows.filter(row=>row.input.context==='workflow'&&['Subscription','Organization','User'].includes(row.input.object)).flatMap(row=>row.cells).filter(row=>row.status==='ok').length;assert.equal(workflowCells,6);
const original=read(`${base}/saved-original/report.json`),successor=read(`${base}/saved-successor/report.json`);
const hashes=report=>Object.fromEntries(report.rows.map(row=>[row.schema+'.json',row.input.sha256]));
assert.deepEqual(hashes(original),hashes(read('artifacts/product-reality/sprint-186/m05/reachability/report.json')));
const adoption=read('artifacts/product-reality/sprint-188/m01/adoption.json');assert.deepEqual(hashes(successor),Object.fromEntries(Object.entries(adoption.after).filter(([file])=>file!=='_index.json')));
assert.equal(original.reachable,15);assert.equal(successor.reachable,16);
const liveRoot='/Users/systemsystems/portfolio/Design-Tools/OODS-Forge/packages/mcp-server/.oods/schemas';
const live=Object.fromEntries(fs.readdirSync(liveRoot).filter(file=>file.endsWith('.json')).sort().map(file=>[file,hash(fs.readFileSync(path.join(liveRoot,file)))]));assert.deepEqual(live,adoption.after);
fs.writeFileSync(`${base}/saved-compatibility.json`,JSON.stringify({sourceHead:after.head,originalInputsUnchanged:true,successorInputsUnchanged:true,liveStoreFiles:Object.keys(live).length,changedLiveFiles:0,liveStore:{path:liveRoot,hashes:live},readOnly:true},null,2)+'\n');
fs.writeFileSync(`${base}/schema-movement-observed.json`,JSON.stringify({head:after.head,statusChanges,workflowCells,changes},null,2)+'\n');
console.log(`77/77,154/154; workflows ${workflowCells}/6; ${changes.length} changed framework artifacts for attribution; original15/16, successor16/16, 17 live hashes unchanged.`);

File: artifacts/product-reality/sprint-197/m07/preparation/verify-census.py
"""Sprint 197 retained-roster comparison: palette and generated paint move; composed schemas remain unchanged."""
import collections
import hashlib
import json
from pathlib import Path

base = Path('artifacts/product-reality/sprint-197/m07')
read = lambda file: json.loads(Path(file).read_text())
sha = lambda data: hashlib.sha256(data).hexdigest()
before = read('artifacts/product-reality/sprint-196/m07/component-census/report.json')
after = read(base / 'component-census/report.json')
classes = {row['id']: row['proposedClassification'] for row in read('packages/component-contracts/registry/component-reconciliation.proposed.v2.json')['rows']}

def nodes(node):
    return [node] + [child for item in node.get('children', []) for child in nodes(item)]

def normalized(schema):
    ids = {node['id']: f'node-{index}' for index, node in enumerate(node for screen in schema['screens'] for node in nodes(screen))}
    def clean(value):
        if isinstance(value, dict): return {key: clean(item) for key, item in value.items()}
        if isinstance(value, list): return [clean(item) for item in value]
        return ids.get(value, value) if isinstance(value, str) else value
    return json.dumps(clean(schema), sort_keys=True, separators=(',', ':'), ensure_ascii=False).encode()

assert after['greenTotalSchemas'] == 77 and after['greenTotalCells'] == 154
assert len(after['allRows']) == len(before['allRows']) == 77
key = lambda row: (row['input']['object'], row['input']['context'])
old_rows = {key(row): row for row in before['allRows']}
assert len(old_rows) == 77 and set(old_rows) == {key(row) for row in after['allRows']}
observed = []
for row in after['allRows']:
    old = old_rows[key(row)]
    for item in [old, row]:
        assert sha(Path(item['composition']['path']).read_bytes()) == item['composition']['sha256']
    left, right = read(old['composition']['path'])['schema'], read(row['composition']['path'])['schema']
    assert normalized(left) == normalized(right), f'Undeclared composed-schema movement: {key(row)}'
    assert {cell['framework'] for cell in row['cells']} == {'react', 'vue'}
    for cell in row['cells']:
        assert cell['status'] == 'ok' and cell['artifactPresent'] and not cell['errors']
        assert sha(Path(cell['response']['path']).read_bytes()) == cell['response']['sha256']
    counts = collections.Counter(node['component'] for screen in right['screens'] for node in nodes(screen))
    observed.append({'input': row['input'], 'before': old['composition'], 'after': row['composition'],
                     'beforeNormalizedHash': sha(normalized(left)), 'afterNormalizedHash': sha(normalized(right)),
                     'components': [{'component': name, 'count': count, 'proposedClassification': classes[name]} for name, count in sorted(counts.items())],
                     'class': 'unchanged-composed-schema', 'added': [], 'removed': []})
result = {'baselineHead': before['head'], 'head': after['head'], 'executionHead': after['head'],
          'changedSchemas': 0, 'unchangedSchemas': 77, 'generationCells': 154, 'passedGenerationCells': 154,
          'classes': {}, 'objectsGainingCharts': [], 'rows': [], 'allRows': observed, 'unattributedChanges': [],
          'normalization': 'Generated node IDs and exact references become traversal indices. All other composed-schema bytes participate. The explicit original eleven-object roster is the retained comparison; current eighteen-object support is separately measured by runtime240. Palette changes are separately bound by m05 goldens and m07 runtime proof.'}
(base / 'schema-movement.json').write_text(json.dumps(result, indent=2) + '\n')
original = read(base / 'saved-original/report.json'); successor = read(base / 'saved-successor/report.json')
hashes = lambda report: {row['schema'] + '.json': row['input']['sha256'] for row in report['rows']}
assert hashes(original) == hashes(read('artifacts/product-reality/sprint-186/m05/reachability/report.json'))
adoption = read('artifacts/product-reality/sprint-188/m01/adoption.json')
assert hashes(successor) == {name: value for name, value in adoption['after'].items() if name != '_index.json'}
assert original['reachable'] == 15 and successor['reachable'] == 16
live = Path('/Users/systemsystems/portfolio/Design-Tools/OODS-Forge/packages/mcp-server/.oods/schemas')
actual = {file.name: sha(file.read_bytes()) for file in sorted(live.glob('*.json'))}
cohort = Path('artifacts/product-reality/sprint-187/m05/delivery-final/recomposed-store')
cohort_hashes = {file.name: sha(file.read_bytes()) for file in sorted(cohort.glob('*.json'))}
assert cohort_hashes == adoption['after'] and len(cohort_hashes) == 17
before = read(base / 'preparation/live-store-before.json')
assert actual == before['hashes'] and len(actual) == 37
index = read(live / '_index.json'); old_index = read(cohort / '_index.json')
by_name = {row['name']: row for row in index['schemas']}
assert all(by_name.get(row['name']) == row for row in old_index['schemas'])
assert all(actual[name] == value for name, value in cohort_hashes.items() if name != '_index.json')
(base / 'saved-compatibility.json').write_text(json.dumps({'sourceHead': after['head'], 'originalInputsUnchanged': True, 'successorInputsUnchanged': True,
    'cohortFiles': 17, 'changedCohortFiles': 0, 'cohortStore': {'path': str(cohort), 'hashes': cohort_hashes},
    'liveStoreFiles': 37, 'changedLiveFiles': 0, 'liveStore': {'path': str(live), 'hashes': actual}, 'readOnly': True,
    'snapshotRoot': str(base / 'saved-live-snapshot'), 'primaryBaseline': 'preparation/live-store-before.json', 'originalLiveSchemaFilesUnchanged': 16, 'originalIndexEntriesUnchanged': 16,
    'preExistingIndexExpansion': {'updatedAt': index['updatedAt'], 'additionalSchemas': sorted(set(actual) - set(cohort_hashes)),
      'beforeSha256': cohort_hashes['_index.json'], 'currentSha256': actual['_index.json'],
      'qualification': 'The live index grew on 2026-09-12 before Sprint197; its raw hash differs from the Sprint196 receipt. The fixed17-file cohort and sixteen original live schema files remain byte-identical; all original index entries are unchanged. The current37-file live store is read-only and hash-compared before/after this closeout.'}}, indent=2) + '\n')
print('77 unchanged composed schemas,154 generation cells;17 archived cohort files;37 current live files unchanged during closeout.')
